dapp/request_limitation/throttling.go

```diff
@@ -1,52 +1,85 @@
 package request_limitation
 
 import (
-	"sync"
 	"time"
 )
 
 type Throttling struct {
 	concurrency    uint
 	coolDown       time.Duration
 	stack          chan func()
-	lock           sync.Mutex
 	maxQueue       uint
-	inWork         uint
 	queueFullError error
 }
 
+func (t *Throttling) Close() error {
+	close(t.stack)
+	return nil
+}
+
 // create new throttling request limitation
 func NewThrottling(concurrency uint, coolDown time.Duration, maxQueue uint, queueFullError error) *Throttling {
 
 	t := &Throttling{
 		concurrency:    concurrency,
 		coolDown:       coolDown,
 		stack:          make(chan func(), maxQueue),
-		lock:           sync.Mutex{},
 		maxQueue:       maxQueue,
 		queueFullError: queueFullError,
 	}
 
+	inWorkChan := make(chan chan uint)
+	incInWork := make(chan struct{})
+	decInWork := make(chan struct{})
+
+	// state
 	go func() {
+
+		var inWork uint
+
 		for {
-			t.lock.Lock()
-			if t.inWork >= t.concurrency {
-				t.lock.Unlock()
+
+			// exit when stack got closed
+			if t.stack == nil {
+				return
+			}
+
+			select {
+			case respChan := <-inWorkChan:
+				respChan <- inWork
+			case <-incInWork:
+				inWork++
+			case <-decInWork:
+				inWork--
```

Comment:
uint has a default value of 0

Decreasing it directly can lead to an underflow and `inWork` would become `4294967295`

This is likely not an expected behaviour